internal/controller/iprangeclaim_controller.go

```diff
@@ -0,0 +1,323 @@
+	if err != nil {
+		return netboxv1.IpRangeClaimStatus{}, err
+	}
+
+	ipAddresses := []string{}
+	ipAddressesDotDecimal := []string{}
+	for _, ip := range availableIps.Payload {
+		ipAddresses = append(ipAddresses, ip.Address)
+		ipAddressesDotDecimal = append(ipAddressesDotDecimal, strings.Split(ip.Address, "/")[0])
+	}
+
+	return netboxv1.IpRangeClaimStatus{
+		IpRange:                fmt.Sprintf("%s-%s", ipRange.Spec.StartAddress, ipRange.Spec.EndAddress),
+		IpRangeDotDecimal:      fmt.Sprintf("%s-%s", startAddressDotDecimal, endAddressDotDecimal),
+		IpAddresses:            ipAddresses,
+		IpAddressesDotDecimal:  ipAddressesDotDecimal,
+		StartAddress:           ipRange.Spec.StartAddress,
+		StartAddressDotDecimal: startAddressDotDecimal,
+		EndAddress:             ipRange.Spec.EndAddress,
+		EndAddressDotDecimal:   endAddressDotDecimal,
+		IpRangeName:            ipRange.Name,
+		Conditions:             o.Status.Conditions,
+	}, nil
+}
+
+func (r *IpRangeClaimReconciler) restoreOrAssignIpRangeAndSetCondition(ctx context.Context, o *netboxv1.IpRangeClaim) (*models.IpRange, ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	ll, res, err := r.tryLockOnParentPrefix(ctx, o)
+	if err != nil {
+		return nil, res, err
+	}
+
+	h := generateIpRangeRestorationHash(o)
+	ipRangeModel, err := r.NetboxClient.RestoreExistingIpRangeByHash(h)
+	if err != nil {
+		if err := r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeAssignedFalse, corev1.EventTypeWarning, "", err); err != nil {
+			return nil, ctrl.Result{}, err
+		}
+		return nil, ctrl.Result{Requeue: true}, nil
+	}
+
+	if ipRangeModel == nil {
+		// ip range cannot be restored from netbox
+		// assign new available ip range
+		ipRangeModel, err = r.NetboxClient.GetAvailableIpRangeByClaim(
+			&models.IpRangeClaim{
+				ParentPrefix: o.Spec.ParentPrefix,
+				Size:         o.Spec.Size,
+				Metadata: &models.NetboxMetadata{
+					Tenant: o.Spec.Tenant,
+				},
+			},
+		)
+		if err != nil {
+			if err := r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeAssignedFalse, corev1.EventTypeWarning, "", err); err != nil {
+				return nil, ctrl.Result{}, err
+			}
+			return nil, ctrl.Result{Requeue: true}, nil
+		}
+		logger.V(4).Info(fmt.Sprintf("ip range is not reserved in netbox, assigned new ip range: %s-%s", ipRangeModel.StartAddress, ipRangeModel.EndAddress))
+	} else {
+		// reassign reserved ip range from netbox
+
+		// check if the restored ip range has the size requested by the claim
+		availableIpRanges, err := r.NetboxClient.GetAvailableIpAddressesByIpRange(ipRangeModel.Id)
+		if len(availableIpRanges.Payload) != o.Spec.Size {
+			ll.Unlock()
+			err = r.logErrorSetConditionAndCreateEvent(ctx, o, netboxv1.ConditionIpRangeAssignedFalseSizeMissmatch, corev1.EventTypeWarning, "", err)
+			if err != nil {
+				return nil, ctrl.Result{}, err
+			}
+			return nil, ctrl.Result{Requeue: true}, nil
```

Comment:
Do we want to requeue in this case?

IMHO if `len(availableIpRanges.Payload) != o.Spec.Size` is triggered, it would require human intervention to fix it.

Reply:
This case needs human intervention, but with exponential back off there won't be too many retries and the claim could get ready if the conflicting data in NetBox is cleaned up. :)